src/librbd/internal.cc

```diff
@@ -2483,9 +2500,117 @@ int validate_pool(IoCtx &io_ctx, CephContext *cct) {
     return 0;
   }
 
-  int mirror_image_get(ImageCtx *ictx, mirror_image_t *mirror_image) {
+  int mirror_image_promote(ImageCtx *ictx, bool force) {
     CephContext *cct = ictx->cct;
-    ldout(cct, 20) << "mirror_image_get " << ictx << dendl;
+    ldout(cct, 20) << __func__ << ": ictx=" << ictx << ", "
+                   << "force=" << force << dendl;
+
+    int r = validate_mirroring_enabled(ictx);
+    if (r < 0) {
+      return r;
+    }
+
+    std::string mirror_uuid;
+    r = Journal<>::get_tag_owner(ictx, &mirror_uuid);
+    if (r < 0) {
+      lderr(cct) << "failed to determine tag ownership: " << cpp_strerror(r)
+                 << dendl;
+      return r;
+    } else if (mirror_uuid == Journal<>::LOCAL_MIRROR_UUID) {
+      lderr(cct) << "image is already primary" << dendl;
+      return -EINVAL;
+    } else if (mirror_uuid != Journal<>::ORPHAN_MIRROR_UUID && !force) {
+      lderr(cct) << "image is still primary within a remote cluster" << dendl;
+      return -EBUSY;
+    }
+
+    r = Journal<>::allocate_tag(ictx, Journal<>::LOCAL_MIRROR_UUID);
```

Comment:
I guess we might want an interlock with rbd-mirror here so that it doesn't have any i/o in flight to the local image after it's promoted, so it can be used after promotion immediately